# Reject malformed material, user element and load input
#
#   - designvariabless: reject an unknown *DESIGN VARIABLES type
#   - dloads: reject a non-positive element number on *DLOAD
#   - elastics: reject an *ELASTIC card without data lines
#   - usermaterials: require a positive CONSTANTS on *USER MATERIAL
#   - mohrcoulombhardenings: require a preceding *MOHR COULOMB
#   - userelements: require positive NODES, INTEGRATION POINTS, MAXDOF
#   - usersections: require a positive CONSTANTS on *USER SECTION
#   - damageinitiations: reject *DAMAGE INITIATION without CRITERION
#   - initialconditionss: check element nodes before evaluating sigini
#   - calinput: stop when no element was defined
#   - elementcpuload: give one empty range when no element is active
#   - allocation: initialise nconstants for *DAMAGE INITIATION
#   - films: reject a non-positive element number on *FILM
#   - radiates: reject a non-positive element number on *RADIATE
#
# Found by fuzzing ccx with malformed variants of 518 input decks under
# AddressSanitizer, and by review of the code around each crash.

diff --git a/ccx_2.23/src/allocation.f b/ccx_2.23/src/allocation.f
index 18ccea8..f696005 100644
--- a/ccx_2.23/src/allocation.f
+++ b/ccx_2.23/src/allocation.f
@@ -516,6 +516,7 @@
         enddo
       elseif(textpart(1)(1:17).eq.'*DAMAGEINITIATION') then
         ntmatl=0
+        nconstants=0
 !
         do i=2,n
           if(textpart(i)(1:10).eq.'CRITERION=') then
diff --git a/ccx_2.23/src/calinput.f b/ccx_2.23/src/calinput.f
index e7b74d1..1fdbfe4 100644
--- a/ccx_2.23/src/calinput.f
+++ b/ccx_2.23/src/calinput.f
@@ -1622,6 +1622,13 @@ c
         nprint=ii
       endif
 !     
+!     check whether any element was defined
+!     
+      if(ne.eq.0) then
+        write(*,*) '*ERROR in calinput: no elements were defined'
+        call exit(201)
+      endif
+!     
 !     check whether a material was assigned to each active element
 !     
       ierror=0
diff --git a/ccx_2.23/src/damageinitiations.f b/ccx_2.23/src/damageinitiations.f
index f45ae66..8d49d83 100644
--- a/ccx_2.23/src/damageinitiations.f
+++ b/ccx_2.23/src/damageinitiations.f
@@ -43,6 +43,8 @@
          ier=1
          return
       endif
+!
+      itype=0
 !
       do i=2,n
         if(textpart(i)(1:10).eq.'CRITERION=') then
@@ -69,6 +71,14 @@
      &         "*DAMAGE INITIATION%")
         endif
       enddo
+!
+      if(itype.eq.0) then
+        write(*,*) 
+     &       '*ERROR reading *DAMAGE INITIATION: criterion is lacking'
+        call inputerror(inpc,ipoinpc,iline,
+     &       "*DAMAGE INITIATION%",ier)
+        return
+      endif
 !
 !     the damage initiation is stored as a mechanical user material
 !
diff --git a/ccx_2.23/src/designvariabless.f b/ccx_2.23/src/designvariabless.f
index 7f94bb5..9bc7d33 100644
--- a/ccx_2.23/src/designvariabless.f
+++ b/ccx_2.23/src/designvariabless.f
@@ -82,6 +82,15 @@
      &        "*DESIGN VARIABLES%",ier)
          return
       endif
+!
+      if((tieset(1,ntie)(1:10).ne.'COORDINATE').and.
+     &   (tieset(1,ntie)(1:11).ne.'ORIENTATION')) then
+         write(*,*)
+     &'*ERROR reading *DESIGN VARIABLES: unknown type'
+         call inputerror(inpc,ipoinpc,iline,
+     &        "*DESIGN VARIABLES%",ier)
+         return
+      endif
 !
 !     Add "D" at the end of the name of the designvariable keyword
 !      
diff --git a/ccx_2.23/src/dloads.f b/ccx_2.23/src/dloads.f
index 4858458..59b92ce 100644
--- a/ccx_2.23/src/dloads.f
+++ b/ccx_2.23/src/dloads.f
@@ -329,7 +329,7 @@ c     BernhardiEnd
 !     
         read(textpart(1)(1:10),'(i10)',iostat=istat) l
         if(istat.eq.0) then
-          if(l.gt.ne) then
+          if((l.gt.ne).or.(l.lt.1)) then
             write(*,*) '*ERROR reading *DLOAD: element ',l
             write(*,*) '       is not defined'
             ier=1
diff --git a/ccx_2.23/src/elastics.f b/ccx_2.23/src/elastics.f
index 910144b..7a199bc 100644
--- a/ccx_2.23/src/elastics.f
+++ b/ccx_2.23/src/elastics.f
@@ -85,7 +85,14 @@
          do
             call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &           ipoinp,inp,ipoinpc)
-            if((istat.lt.0).or.(key.eq.1)) return
+            if((istat.lt.0).or.(key.eq.1)) then
+               if(ntmat.eq.0) then
+                  write(*,*) 
+     &            '*ERROR reading *ELASTIC: *ELASTIC card without data'
+                  ier=1
+               endif
+               return
+            endif
             ntmat=ntmat+1
             nelcon(2,nmat)=ntmat
             if(ntmat.gt.ntmat_) then
@@ -144,7 +151,14 @@
          do
             call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &           ipoinp,inp,ipoinpc)
-            if((istat.lt.0).or.(key.eq.1)) return
+            if((istat.lt.0).or.(key.eq.1)) then
+               if(ntmat.eq.0) then
+                  write(*,*) 
+     &            '*ERROR reading *ELASTIC: *ELASTIC card without data'
+                  ier=1
+               endif
+               return
+            endif
             ntmat=ntmat+1
             nelcon(2,nmat)=ntmat
             if(ntmat.gt.ntmat_) then
@@ -223,7 +237,14 @@
          do
             call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &           ipoinp,inp,ipoinpc)
-            if((istat.lt.0).or.(key.eq.1)) return
+            if((istat.lt.0).or.(key.eq.1)) then
+               if(ntmat.eq.0) then
+                  write(*,*) 
+     &            '*ERROR reading *ELASTIC: *ELASTIC card without data'
+                  ier=1
+               endif
+               return
+            endif
             ntmat=ntmat+1
             nelcon(2,nmat)=ntmat
             if(ntmat.gt.ntmat_) then
diff --git a/ccx_2.23/src/elementcpuload.c b/ccx_2.23/src/elementcpuload.c
index 32897aa..2a6d3df 100644
--- a/ccx_2.23/src/elementcpuload.c
+++ b/ccx_2.23/src/elementcpuload.c
@@ -42,6 +42,17 @@ void elementcpuload(ITG *neapar,ITG *nebpar,ITG *ne,ITG *ipkon,ITG *num_cpus){
 	    nepar++;
 	}
     }
+    if(nepar==0){
+
+	/* no active elements (e.g. all deleted): one thread with
+           an empty element range */
+
+	*num_cpus=1;
+	neapar[0]=0;
+	nebpar[0]=-1;
+	SFREE(ipar);
+	return;
+    }
     if(nepar<*num_cpus) *num_cpus=nepar;
 
     /* dividing the element number range into num_cpus equal numbers of 
diff --git a/ccx_2.23/src/films.f b/ccx_2.23/src/films.f
index c4f1797..4522098 100644
--- a/ccx_2.23/src/films.f
+++ b/ccx_2.23/src/films.f
@@ -270,7 +270,7 @@ c     call reorderampl(amname,namta,nam)
 !     
         read(textpart(1)(1:10),'(i10)',iostat=istat) l
         if(istat.eq.0) then
-          if(l.gt.ne) then
+          if((l.gt.ne).or.(l.lt.1)) then
             write(*,*) '*ERROR reading *FILM: element ',l
             write(*,*) '       is not defined'
             ier=1
diff --git a/ccx_2.23/src/initialconditionss.f b/ccx_2.23/src/initialconditionss.f
index a58303a..ea06535 100644
--- a/ccx_2.23/src/initialconditionss.f
+++ b/ccx_2.23/src/initialconditionss.f
@@ -231,6 +231,13 @@ c                    t1(k)=temperature
 !     
                 do j=1,nope
                   konl(j)=kon(indexe+j)
+                  if((konl(j).lt.1).or.(konl(j).gt.nk)) then
+                    write(*,*) '*ERROR reading *INITIAL CONDITIONS:'
+                    write(*,*) '       element ',i,' contains node ',
+     &                   konl(j),' which does not exist'
+                    ier=1
+                    return
+                  endif
                   do k=1,3
                     xl(k,j)=co(k,konl(j))
                   enddo
diff --git a/ccx_2.23/src/mohrcoulombhardenings.f b/ccx_2.23/src/mohrcoulombhardenings.f
index 2748375..e48a492 100644
--- a/ccx_2.23/src/mohrcoulombhardenings.f
+++ b/ccx_2.23/src/mohrcoulombhardenings.f
@@ -56,7 +56,15 @@
         ier=1
         return
       endif
-!     
+!
+      if(nelcon(1,nmat).ne.-53) then
+        write(*,*) '*ERROR reading *MOHR COULOMB HARDENING:'
+        write(*,*) '       *MOHR COULOMB HARDENING should be preceded'
+        write(*,*) '       by a *MOHR COULOMB card'
+        ier=1
+        return
+      endif
+!
       iperturb(1)=3
 !     
       do i=2,n
diff --git a/ccx_2.23/src/radiates.f b/ccx_2.23/src/radiates.f
index 82f6f39..f7c0812 100644
--- a/ccx_2.23/src/radiates.f
+++ b/ccx_2.23/src/radiates.f
@@ -296,7 +296,7 @@ c     call reorderampl(amname,namta,nam)
 !     
         read(textpart(1)(1:10),'(i10)',iostat=istat) l
         if(istat.eq.0) then
-          if(l.gt.ne) then
+          if((l.gt.ne).or.(l.lt.1)) then
             write(*,*) '*ERROR reading *RADIATE: element ',l
             write(*,*) '       is not defined'
             ier=1
diff --git a/ccx_2.23/src/userelements.f b/ccx_2.23/src/userelements.f
index 2fe3cb7..c06eb6a 100644
--- a/ccx_2.23/src/userelements.f
+++ b/ccx_2.23/src/userelements.f
@@ -31,6 +31,9 @@
      &  inl,ipol
 !
       four=4
+      nodes=0
+      intpoints=0
+      maxdof=0
 !
       do i=2,n
          if(textpart(i)(1:6).eq.'TYPE=U') then
@@ -63,6 +66,29 @@
       enddo
 !
 !     check range
+!
+      if(intpoints.lt.1) then
+         write(*,*) '*ERROR reading *USER ELEMENT'
+         write(*,*) '       number of integration points ',intpoints,
+     &       ' is not positive'
+         ier=1
+         return
+      endif
+!
+      if(maxdof.lt.1) then
+         write(*,*) '*ERROR reading *USER ELEMENT'
+         write(*,*) '       highest degree of freedom ',maxdof,
+     &       ' is not positive'
+         ier=1
+         return
+      endif
+!
+      if(nodes.lt.1) then
+         write(*,*) '*ERROR reading *USER ELEMENT'
+         write(*,*) '       number of nodes ',nodes,' is not positive'
+         ier=1
+         return
+      endif
 !
       if(intpoints.gt.255) then
          write(*,*) '*ERROR reading *USER ELEMENT'
diff --git a/ccx_2.23/src/usermaterials.f b/ccx_2.23/src/usermaterials.f
index 49fbe94..b7d27af 100644
--- a/ccx_2.23/src/usermaterials.f
+++ b/ccx_2.23/src/usermaterials.f
@@ -54,6 +54,7 @@
       endif
 !
       imech=1
+      nconstants=0
 !
       do i=2,n
          if(textpart(i)(1:10).eq.'CONSTANTS=') then
@@ -74,6 +75,15 @@
      &"*USER MATERIAL%")
          endif
       enddo
+!
+      if(nconstants.le.0) then
+         write(*,*) 
+     &   '*ERROR reading *USER MATERIAL: the number of constants'
+         write(*,*) '       (parameter CONSTANTS) must be positive'
+         call inputerror(inpc,ipoinpc,iline,
+     &        "*USER MATERIAL%",ier)
+         return
+      endif
 !
       if(imech.eq.1) then
 !
diff --git a/ccx_2.23/src/usersections.f b/ccx_2.23/src/usersections.f
index 2b9a1e9..4f4d0b2 100644
--- a/ccx_2.23/src/usersections.f
+++ b/ccx_2.23/src/usersections.f
@@ -47,6 +47,8 @@
         ier=1
         return
       endif
+!     
+      ndprop=0
 !     
       do i=2,n
         if(textpart(i)(1:9).eq.'MATERIAL=') then
@@ -74,6 +76,18 @@
         endif
       enddo
 !     
+!     the user elements take their section properties from field
+!     prop, which is only allocated if properties are defined
+!     
+      if(ndprop.le.0) then
+        write(*,*) 
+     &       '*ERROR reading *USER SECTION: the number of constants'
+        write(*,*) '       (parameter CONSTANTS) must be positive'
+        call inputerror(inpc,ipoinpc,iline,
+     &       "*USER SECTION%",ier)
+        return
+      endif
+!     
 !     check for the existence of the set and the material
 !     
       do i=1,nmat
